e2e/auto_complete_test.go

```diff
@@ -0,0 +1,238 @@
+			arg:           "",
+			// local file completions are prepared by the shell
+			expected: []string{},
+			shell:    "/bin/bash",
+		},
+		{
+			name:          "cat complete empty string",
+			precedingArgs: []string{"cat"},
+			arg:           "",
+			expected:      []string{"s3://" + bucket + "/"},
+			shell:         "/bin/pwsh",
+		},
+		{
+			name:          "mb complete empty string",
+			precedingArgs: []string{"mb"},
+			arg:           "",
+			expected:      []string{"s3://"},
+			shell:         "/bin/pwsh",
+		},
+		{
+			name:          "mb complete bucket",
+			precedingArgs: []string{"mb"},
+			arg:           "s3://bu",
+			expected:      []string{"s3://bu"},
+			shell:         "/bin/pwsh",
+		},
+		{
+			name:          "rb complete empty string",
+			precedingArgs: []string{"rb"},
+			arg:           "",
+			expected:      []string{"s3://" + bucket + "/"},
+			shell:         "/bin/pwsh",
+		},
+		{
+			name:          "rb should not complete keys string",
+			precedingArgs: []string{"rb"},
+			arg:           "s3://" + bucket + "/f",
+			expected:      []string{"s3://" + bucket + "/"},
+			remoteFiles: []string{
+				"file.txt",
+				"fdir/child.txt",
+			},
+			shell: "/bin/pwsh",
+		},
+		{
+			name:          "select complete empty string",
+			precedingArgs: []string{"select"},
+			arg:           "",
+			expected:      []string{"s3://" + bucket + "/"},
+			shell:         "/bin/pwsh",
+		},
+		{
+			name:          "cp complete bucket names in pwsh",
+			precedingArgs: []string{"cp"},
+			arg:           "s3://",
+			expected:      []string{"s3://" + bucket + "/"},
+			shell:         "/bin/pwsh",
+		},
+		{
+			name:          "cp complete bucket names in zsh",
+			precedingArgs: []string{"cp"},
+			arg:           "s3://",
+			expected:      []string{"s3\\://" + bucket + "/"},
+			shell:         "/bin/zsh",
+		},
+		{
+			name:          "cp complete bucket names in bash",
+			precedingArgs: []string{"cp"},
+			arg:           "s3://",
+			expected:      []string{"s3://" + bucket + "/", "//" + bucket + "/"},
+			shell:         "/bin/bash",
+		},
+		{
+			name:          "cp complete bucket keys pwsh",
+			precedingArgs: []string{"cp"},
+			arg:           "s3://" + bucket + "/",
+			remoteFiles: []string{
+				"file0.txt",
+				"file1.txt",
+				"filedir/child.txt",
+				"dir/child.txt",
+			},
+			expected: keysToS3URL("s3://", bucket,
+				"file0.txt",
+				"file1.txt",
+				"filedir/",
+				"dir/",
+			),
+			shell: "/bin/pwsh",
+		},
+		{
+			name:          "cp complete bucket keys bash",
+			precedingArgs: []string{"cp"},
+			arg:           "s3://" + bucket + "/",
+			remoteFiles: []string{
+				"file0.txt",
+				"file1.txt",
+				"filedir/child.txt",
+				"dir/child.txt",
+			},
+			expected: append(
+				keysToS3URL("s3://", bucket,
+					"file0.txt",
+					"file1.txt",
+					"filedir/",
+					"dir/"),
+				keysToS3URL("//", bucket,
+					"file0.txt",
+					"file1.txt",
+					"filedir/",
+					"dir/")...),
```

Comment:
You may just repeat yourself. More explicit the tests more easier to read IMHO, you may just focus on what the tested unit does instead decoding how inputs/outputs are prepared.